### ISSUE TITLE
NoClassDefFoundError with JDK 11

### ISSUE THREAD

**Issue by matthjes** [cdi-unit/cdi-unit#193]
Hi,
we are currently migrating a project from JDK 8 to JDK 11 and all our integration tests fail with the same message:
```
java.lang.NoClassDefFoundError: com/sun/msv/reader/util/IgnoreController

	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:800)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:698)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:621)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:579)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at org.jglue.cdiunit.internal.WeldTestUrlDeployment.loadClass(WeldTestUrlDeployment.java:287)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.jglue.cdiunit.internal.WeldTestUrlDeployment.<init>(WeldTestUrlDeployment.java:187)
	at org.jglue.cdiunit.internal.Weld11TestUrlDeployment.<init>(Weld11TestUrlDeployment.java:28)
	at com.bmw.swhrl.test.util.HrlRunnerClass$1.createDeployment(HrlRunnerClass.java:79)
	at org.jboss.weld.environment.se.Weld.initialize(Weld.java:764)
	at com.bmw.swhrl.test.util.HrlRunnerClass.initializeWeld(HrlRunnerClass.java:87)
	at com.bmw.swhrl.test.util.HrlRunnerClass.classBlock(HrlRunnerClass.java:63)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:362)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: java.lang.ClassNotFoundException: com.sun.msv.reader.util.IgnoreController
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 29 more
```
I've tried to add jaxb-libs to our dependencies, with no avail.

**Comment by ikysil** [comment]
Hi @matthjes,

Thank you for the report.

The stack trace is not enough to do the investigation, given the additional JUnit 4 runners used - see `com.bmw.swhrl.test.util.HrlRunnerClass`.

Could you please create a reproducer project at GitHub?

What JVM flags are you using when running the tests?

We had to add `<argLine>--add-opens java.base/java.lang=ALL-UNNAMED</argLine>` configuration for `maven-surefire-plugin` in our projects while migrating to Java 11 to enable tests execution.

**Comment by matthjes** [comment]
Uff, a reproducer project...this is not possible as the application is huge and it's not trivial to extract just some part of it into an extra project as there are a lot of dependencies, database calls etc. 

I've added the flag, still the same problem. When I add 
```
<dependency>
  <groupId>com.sun.xml.bind</groupId>
  <artifactId>jaxb-libs</artifactId>
  <version>1.0.6</version>
</dependency>
```
the class is found...but now it's asking for `java.lang.NoClassDefFoundError: org/osgi/framework/BundleActivator`...ok, I add 
```
<dependency>
  <groupId>org.osgi</groupId>
  <artifactId>org.osgi.core</artifactId>
  <version>6.0.0</version>
</dependency>
```
and the class is found...but now it's asking for `java.lang.NoClassDefFoundError: com/oracle/wls/shaded/org/apache/xml/serializer/DOMSerializer`...and I'm lost. I have no idea what component requires these classes.

**Comment by ikysil** [comment]
> but now it's asking for `java.lang.NoClassDefFoundError: com/oracle/wls/shaded/org/apache/xml/serializer/DOMSerializer`...and I'm lost. I have no idea what component requires these classes.

In this particular case, `com/oracle/wls` points to Oracle's WebLogic application server.

I agree, it is not easy.

What version of cdi-unit are you using?
Could you reproduce this issue with the `4.1.1-SNAPSHOT` version built from the master branch?

**Comment by matthjes** [comment]
Hi,
I've looked deeper into this and it turned out that there were some weird things going on with class loading. The error occurred when I added a specific class via `@AdditionalClasspaths`. This loaded more project classes then expected, which somehow triggered the loading some classes inside `payara-embedded-all` dependency which then lead to this error.

I've solved this by moving the class into `@AdditionalPackages`, which does not seem to trigger the extensive class loading.
